File: go/aoc2022/day6/src/main.zig
const std = @import("std");

fn solve(allocator: std.mem.Allocator, line: []u8, num: usize) !void {
    var m = std.ArrayList(u8).init(allocator);
    defer m.deinit();

    for (line) |c, i| {
        try m.insert(0, c);

        if (m.items.len == num) {
            var duplicate = false;
            for (m.items) |v, j| {
                var k: usize = j + 1;
                while (k < m.items.len) : (k += 1) {
                    duplicate = duplicate or m.items[k] == v;
                }
            }

            if (!duplicate) {
                std.debug.print("{d}\n", .{i + 1});
                return;
            }

            _ = m.pop();
        }
    }
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    var allocator = arena.allocator();
    var buf = try allocator.alloc(u8, 5000);
    defer allocator.free(buf);

    var cwd = std.fs.cwd();
    const file = try cwd.readFile("in", buf);

    try solve(allocator, file, 4);
    try solve(allocator, file, 14);
}

File: go/aoc2022/day6/main.py

def one(list: str, n: int):
    
    arr = []
    for i, c in enumerate(list):
        arr.insert(0, c)

        duplicate = False
        for k, a in enumerate(arr):
            for j in range(k+1, len(arr)):
                if (arr[j] == a):
                    duplicate = True
                    break

        if len(arr) == n:
            if duplicate == False:
                print(i+1) 
                return
            else:
                arr.pop()
         
            

if __name__ == "__main__":
    s = open("in").read()

    one(s, 4)
    one(s, 14)
